---
title: "ITS2_Isopora_CS_Genus_Symbioant"
author: "Monique White"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Libraries
```{r, warning = FALSE, echo = FALSE, message=FALSE}
library(ggsci)
library(tidyverse)
library(dplyr)
library(forcats)
library(reshape2)
library(stringr)
library(tidyr)
library(tibble)
library(sangerseqR)
library(DECIPHER)
library(Biostrings)
library(phangorn)
library(ape)
library(ggplot2)
library(ggtree)
library(patchwork)
library(bioseq)
library(kmer)
library(GUniFrac)
library(seqinr)
library(vegan)
library(corrplot)
library(ggrepel)

library(ggmsa)
library(dendextend)
library(usedist)
```


```{r}
seqs_iso <- read_tsv("seqs.abs.Isopora.txt") %>% 
    filter(!(is.na(sample_name))) %>%
  dplyr::select(sample_name, `A1`:`G3ai`) %>% 
  mutate(sample_name = as.factor(sample_name)) %>% 
  filter(!grepl("-20", sample_name) & !grepl("b", sample_name))

seqs_iso
```

#Load metadata
Note the metadata requires the imput of the CoralSea2024_Metadata.xlsx. you need to change the path for where the file lives on your in individual computer.

```{r}

library(readxl)
CoralSea2024_Metadata <- read_excel("~/AIMS@JCU/CoralSea2024_Metadata.xlsx")

meta_iso <- CoralSea2024_Metadata %>% 
  rename("VialID" = "sample_name",
         "Depth (m)"= "Depth") #renamed to match the Acropora metadata that was used in ITS2 analysis

meta_iso <- meta_iso %>% 
  select(sample_name, Reference, Reef, Site,Date, Bag, Species,Depth, `Bleach score`, Vial, Year) %>% 
  filter(Species == "Isopora",
         Year != "20")
meta_iso
```

### Load Custom Functions
```{r}
read_fasta_df <- function (file = "") {
  fasta <- readLines(file)
  ind <- grep(">", fasta)
  s <- data.frame(ind = ind, from = ind + 1, to = c((ind - 
    1)[-1], length(fasta)))
  seqs <- rep(NA, length(ind))
  for (i in 1:length(ind)) {
    seqs[i] <- paste(fasta[s$from[i]:s$to[i]], collapse = "")
  }
  tib <- tibble(label = gsub(">", "", fasta[ind]), sequence = seqs)
  return(tib)
}

write_fasta_df <- function (data, filename) 
{
    fastaLines = c()
    for (rowNum in 1:nrow(data)) {
        fastaLines = c(fastaLines, as.character(paste(">", 
            data[rowNum, "label"], sep = "")))
        fastaLines = c(fastaLines, as.character(data[rowNum, 
            "sequence"]))
    }
    fileConn <- file(filename)
    writeLines(fastaLines, fileConn)
    close(fileConn)
}

dna_to_DNAbin <- function (dna){
  DNAbin <- as_DNAbin(dna)
  names(DNAbin) <- names(dna)
  return(DNAbin)
}
dna_to_DNAStringset <- function(x) 
{
    bioseq:::check_dna(x)
    DNAstr <- DNAStringSet(paste(x))
    names(DNAstr) <- names(x)
    return(DNAstr)
}

DNAStringSet_to_dna <- function(x){
    x_dna <- as_dna(paste(x))
    names(x_dna) <- names(x)
    res <- tibble(label = names(x), sequence = x_dna)
    return(res)
}

# Convert DNAstringset to DNAbin
DNAStringSet_to_DNAbin <- function(DNAStringSet){
  DNAbin <- as.DNAbin(DNAStringSet)
  return(DNAbin)
}

# https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2315-y
palette <- c("A" = "#46ff2d", 
             "G" = "#ffae01", 
             "C" = "#f24641", 
             "T" = "#4294fa", 
             "K" = "#8b4816",
             "M" = "#83831f",
             "R" = "#ffff81",
             "S" = "#ff9d80",
             "Y" = "#e381f2",
             "W" = "#80fff2",
             "V" = "#fde4b8",
             "B" = "#f9c1bf",
             "H" = "#c0d9f9",
             "D" = "#c7ffba",
             "U" = "#8989fb",
             "N" = "black", 
             "-" = "white",
             "+" = "White")


pal_df <- data.frame(names = names(palette), col = palette)
```

### Combining sequence/metadata 
```{r}
# Convert to long format
seqs_long <- seqs_iso %>%
  filter(!is.na(sample_name)) %>%
  select(sample_name, `A1`:`G3ai`) %>%
  pivot_longer(`A1`:`G3ai`) %>%
  filter(value > 0) %>% # Remove zero values
  left_join(., meta_iso)
```

```{r}
#filter out non-profile sequences
load("~/AIMS@JCU/CoralSeaSymbioants/ITS2_Isopora_CS_2024/symbioant_iso_all.RData")

seq_data_A <- symbioant_iso_all %>% 
  filter(!str_detect(name, "non")) %>% 
  filter(!str_detect(name, "p_")) %>% 
  filter(str_sub(name, 1, 1) == "A" | str_detect(name, "_A"))

seq_data_C <- symbioant_iso_all %>% 
  filter(!str_detect(name, "non")) %>% 
  filter(!str_detect(name, "p_")) %>%
  filter(str_sub(name, 1, 1) == "C" | str_detect(name, "_C"))

seq_data_D <- symbioant_iso_all %>% 
  filter(!str_detect(name, "non")) %>% 
  filter(!str_detect(name, "p_")) %>%
  filter(str_sub(name, 1, 1) == "D" | str_detect(name, "_D"))
```


### Filter by sumbioant genus

## Symbiodinium
```{r}
seqs_long_A <- seqs_long %>% 
  filter(str_sub(name, 1, 1) == "A" | str_detect(name, "_A"))
```

```{r}
san_check_A <- seqs_long_A %>% 
  group_by(sample_name) %>%
  summarise(total = sum(value))

#List of samples that have sufficient depth
keepers_ss_A <- san_check_A%>%
  filter(total > 1500) #from 81 to 2

# Filter out the failed samples
seqs_long_A <- seqs_long_A %>%
  filter(sample_name %in% keepers_ss_C$sample_name) %>%
  group_by(sample_name) %>%
  mutate(value_rel = value/sum(value)) %>% # Convert to relative abundance
  ungroup() %>%
  mutate(name = as.factor(name)) # Make sample names a factor
```

```{r}
fasta_iso_A <- read_fasta_df("seqs.isopora.fasta") %>% 
   filter(label %in% seqs_long_A$name) %>%   #only keeping DNA seqs that appear in seqs_long subset 
  deframe() %>%
  as_dna()
```


```{r}
#creating the tree
kdist_iso_A <- fasta_iso_A %>%
  dna_to_DNAbin() %>%
  kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()
```

```{r}
k_tree_iso_A <- kdist_iso_A %>% phangorn::upgma()

k_tree_iso_A #108
```
```{r}
seqs_wide_iso_A <- seq_data_A %>%
  dplyr::select(sample_name, name, value) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  tibble::column_to_rownames(var = "sample_name")
head(seqs_wide_iso_A) #108
```

```{r}
k_unidist_iso_A <- GUniFrac(seqs_wide_iso_A, k_tree_iso_A)   #GUniFrac calculates all the distances 
k_unidist_iso_A <- k_unidist_iso_A$unifracs


 # GUniFrac with alpha 0.5 (i.e Generalised UniFrac Distance)
du_iso_A <- k_unidist_iso_A[, , "d_0.5"]   
dist_iso_A <- as.dist(du_iso_A, diag = FALSE)
```
### PCoa
```{r}
pcoa_iso_A <- cmdscale(dist_iso_A, eig = TRUE)    #this is doing 'cmds' or classic multidimensional scaling

#outlier_samples <- XX

ordiplot(pcoa_iso_A, display = 'sites', type = 'text')
barplot (pcoa_iso_A$eig, names = paste ('PCoA', 1:187), las = 5, ylab = 'eigenvalues')

# Custom x and y labels
xlab = paste("Axis 1 (X%)")
ylab = paste("Axis 2 (X%)")

MDSxy.iso_A <- data.frame(pcoa_iso_A$points) %>% 
  rownames_to_column(var = "sample_name") %>% 
  left_join(., meta_iso)# %>% 
 #  mutate(catBleaching = as.factor(catBleaching)) %>% 
 # filter(!(Vial %in% outlier_samples)) 

isoPCA_A <- MDSxy.iso_A %>%
  filter(Reef != "NA") %>% 
 # mutate(Reef = fct_relevel(Reef, reef_order)) %>%
  ggplot(aes(X1, X2, fill = Reef)) + 
  geom_point(alpha = 1, shape = 21, size = 3) + 
  scale_fill_viridis_d(option = "magma", direction = -1)+ 
#  scale_fill_manual(values = reef_pal) #+
 theme(legend.position = "none", aspect.ratio = 1, text = element_text(size = 15))

isoPCA_A
```


#Cladicopium

```{r}
seqs_long_C <- seqs_long %>% 
  filter(str_sub(name, 1, 1) == "C" | str_detect(name, "_C"))
```

```{r}
san_check_C <- seqs_long_C %>% 
  group_by(sample_name) %>%
  summarise(total = sum(value))

#List of samples that have sufficient depth
keepers_ss_C <- san_check_C%>%
  filter(total > 1500) #from 187 to 187

# Filter out the failed samples
seqs_long_C <- seqs_long_C %>%
  filter(sample_name %in% keepers_ss_C$sample_name) %>%
  group_by(sample_name) %>%
  mutate(value_rel = value/sum(value)) %>% # Convert to relative abundance
  ungroup() %>%
  mutate(name = as.factor(name)) # Make sample names a factor
```

```{r}
fasta_iso_C <- read_fasta_df("seqs.isopora.fasta") %>% 
   filter(label %in% seqs_long_C$name) %>%   #only keeping DNA seqs that appear in seqs_long subset 
  deframe() %>%
  as_dna()
```


```{r}
#creating the tree
kdist_iso_C <- fasta_iso_C %>%
  dna_to_DNAbin() %>%
  kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()
```

```{r}
k_tree_iso_C <- kdist_iso_C %>% phangorn::upgma()

k_tree_iso_C
```
```{r}
seqs_wide_iso_C <- seq_data_C %>%
  dplyr::select(sample_name, name, value) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  tibble::column_to_rownames(var = "sample_name")
head(seqs_wide_iso_C)
```


```{r}
k_unidist_iso_C <- GUniFrac(seqs_wide_iso_C, k_tree_iso_C)   #GUniFrac calculates all the distances 
k_unidist_iso_C <- k_unidist_iso_C$unifracs


 # GUniFrac with alpha 0.5 (i.e Generalised UniFrac Distance)
du_iso_C <- k_unidist_iso_C[, , "d_0.5"]   
dist_iso_C <- as.dist(du_iso_C, diag = FALSE)
```
### PCoa
```{r}
pcoa_iso_C <- cmdscale(dist_iso_C, eig = TRUE)    #this is doing 'cmds' or classic multidimensional scaling

#outlier_samples <- XX

ordiplot(pcoa_iso_C, display = 'sites', type = 'text')
barplot (pcoa_iso_C$eig, names = paste ('PCoA', 1:187), las = 5, ylab = 'eigenvalues')

# Custom x and y labels
xlab = paste("Axis 1 (X%)")
ylab = paste("Axis 2 (X%)")

MDSxy.iso_C <- data.frame(pcoa_iso_C$points) %>% 
  rownames_to_column(var = "sample_name") %>% 
  left_join(., meta_iso)# %>% 
 #  mutate(catBleaching = as.factor(catBleaching)) %>% 
 # filter(!(Vial %in% outlier_samples)) 

isoPCA_C <- MDSxy.iso_C %>%
  filter(Reef != "NA") %>% 
 # mutate(Reef = fct_relevel(Reef, reef_order)) %>%
  ggplot(aes(X1, X2, fill = Reef)) + 
  geom_point(alpha = 1, shape = 21, size = 3) + 
  scale_fill_viridis_d(option = "magma", direction = -1)+ 
#  scale_fill_manual(values = reef_pal) #+
 theme(legend.position = "none", aspect.ratio = 1, text = element_text(size = 15))

isoPCA_C
```

### Procrustes
```{r}
load("~/AIMS@JCU/CoralSeaSymbioants/ITS2_Isopora_CS_2024/Isopora/ind_coords_iso.RData")

iso_seqs_host <- ind_coords_iso %>% 
  mutate("sample_name" = VialID)

#code for defining specific colours per reef for procrustes plots later on
reef_pal2 <- c("#ac1917", "#e54a50", "#b75420", "#ea8553", "#c0982b", "#eec76b", "#768b45", "#a9be77", "#B2DFDB", "#4eaaa6", "#00798C")

reef_names <- c("Elizabeth", "Middleton", "Cato", "Wreck", "Kenn", "Mellish", "Marion","Frederick", "Saumarez", "Madgelaine", "NA")

names(reef_pal2) <- reef_names
```

```{r}
#here is the subset with the data that matches the host data
shared_seqs_iso_C <- inner_join(iso_seqs_host, seq_data_C, by = "sample_name")
str(shared_seqs_iso_C)


#shared_seqs_iso_C <- shared_seqs_iso_C %>%   select(sample_name)
```

Host tree

```{r}
load("~/AIMS@JCU/CoralSeaSymbioants/ITS2_Isopora_CS_2024/Isopora/iso_gen.RData") #load host data

library(adegenet) #install required package
library(poppr)


# Check your individual names in iso_gen
ind_names_host <- indNames(iso_gen)

# Check the sample names you want to keep
sample_names_C <- shared_seqs_iso_C$sample_name

# Subset the genind object to retain only matching individuals
iso_gen_subset_C <- iso_gen[ind_names_host %in% shared_seqs_iso_C$sample_name, ]

length(indNames(iso_gen))          # Before 189
length(indNames(iso_gen_subset_C))   # After 188

#Compute the pairwise distances between multivariate observations; used on raw allele frequencies.
host_shared_dist <- dist(iso_gen_subset_C, method = "euclidean") #calculate euclidean distance 

# convert the distance matrix to a dist object
host_dist_matrix <- as.dist(host_shared_dist)

length(host_dist_matrix) #17578
length(dist_iso_C) #17391
```

# Durusdinium
```{r}
seqs_long_D <- seqs_long %>% 
  filter(str_sub(name, 1, 1) == "D" | str_detect(name, "_D"))
```


```{r}
san_check_D <- seqs_long_D %>% 
  group_by(sample_name) %>%
  summarise(total = sum(value))

#List of samples that have sufficient depth
keepers_ss_D <- san_check_D%>%
  filter(total > 1500) #from 39 to 21

# Filter out the failed samples
seqs_long_D <- seqs_long_D %>%
  filter(sample_name %in% keepers_ss_D$sample_name) %>%
  group_by(sample_name) %>%
  mutate(value_rel = value/sum(value)) %>% # Convert to relative abundance
  ungroup() %>%
  mutate(name = as.factor(name)) # Make sample names a factor
```

```{r}
fasta_iso_D <- read_fasta_df("seqs.isopora.fasta") %>% 
   filter(label %in% seqs_long_D$name) %>%   #only keeping DNA seqs that appear in seqs_long subset 
  deframe() %>%
  as_dna()
```



```{r}
#creating the tree
kdist_iso_D <- fasta_iso_D %>%
  dna_to_DNAbin() %>%
  kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()
```

```{r}
k_tree_iso_D <- kdist_iso_D %>% phangorn::upgma()

k_tree_iso_D #why does the tree have 59 but the following table has 62?
```

```{r}
seqs_wide_iso_D <- seq_data_D %>%
  dplyr::select(sample_name, name, value) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  tibble::column_to_rownames(var = "sample_name")
head(seqs_wide_iso_D) #62
```


```{r}
k_unidist_iso_D <- GUniFrac(seqs_wide_iso_D, k_tree_iso_D)   #GUniFrac calculates all the distances 
k_unidist_iso_D <- k_unidist_iso_D$unifracs


 # GUniFrac with alpha 0.5 (i.e Generalised UniFrac Distance)
du_iso_D <- k_unidist_iso_D[, , "d_0.5"]   
dist_iso_D <- as.dist(du_iso_D, diag = FALSE)
```


